packages/plugin-compiler/src/main/kotlin/io/realm/kotlin/compiler/RealmModelDefaultMethodGeneration.kt

```diff
@@ -0,0 +1,123 @@
+package io.realm.kotlin.compiler
+
+import org.jetbrains.kotlin.backend.common.extensions.IrPluginContext
+import org.jetbrains.kotlin.ir.builders.irGet
+import org.jetbrains.kotlin.ir.builders.irGetObject
+import org.jetbrains.kotlin.ir.builders.irReturn
+import org.jetbrains.kotlin.ir.declarations.IrClass
+import org.jetbrains.kotlin.ir.declarations.IrProperty
+import org.jetbrains.kotlin.ir.declarations.IrSimpleFunction
+import org.jetbrains.kotlin.ir.expressions.impl.IrCallImpl
+import org.jetbrains.kotlin.ir.types.IrType
+import org.jetbrains.kotlin.ir.util.functions
+import org.jetbrains.kotlin.name.Name
+
+/**
+ * Class responsible for adding Realm specific logic to the default object methods like:
+ * - toString()
+ * - hashCode()
+ * - equals()
+ *
+ * WARNING: The current logic in here does not work well with incremental compilation. The reason
+ * is that we check if these methods are "empty" before filling them out, and during incremental
+ * compilation they already have content, and since all of these methods are using inlined
+ * methods they will not pick up changes in the RealmObjectHelper.
+ *
+ * This should only impact us as SDK developers though, but it does mean that changes to
+ * RealmObjectHelper methods will require a clean build to take effect.
+ */
+class RealmModelDefaultMethodGeneration(private val pluginContext: IrPluginContext) {
```

Comment:
Maybe a bit late 😬 But think we could also just have added the default implementation to `RealmObjectInternal` as we previously did with other methods in https://github.com/realm/realm-kotlin/pull/395/files#diff-038fb0f3bf2933f8d86698d2db20de0a0cc1a6406f3bffb49a776db727a92979R39 and then add their fake overrides as done in https://github.com/realm/realm-kotlin/pull/395/files#diff-c7a212a9e2ede847259ae7dc8a0f8ba5872c8238147dbd94d71c12754bcd5b13R65 if they weren't defined by the user.

Reply:
Hmm, I don't know if default methods can override equals and hashCode, but I'll try to experiment with it.

Reply:
Looks like it isn't possible, which after thinking about it, does make sense:

![image](https://github.com/realm/realm-kotlin/assets/406066/a9117715-3a9a-486f-99a5-64a12d983d72)